gradio/blocks.py

```diff
@@ -251,6 +254,17 @@ def __init__(
         self.auth = None
         self.dev_mode = True
         self.app_id = random.getrandbits(64)
+        self.thread_limit_set = False
+
+    async def set_thread_limit(self):
+        if self.thread_limit_set:
+            return
+        self.thread_limit_set = True
+        if self.max_threads is None:
+            return
+        limiter = CapacityLimiter(total_tokens=self.max_threads)
+        prev_run_in_thread = anyio.to_thread.run_sync
+        anyio.to_thread.run_sync = partial(prev_run_in_thread, limiter=limiter)
```

Comment:
Let me clarify this, as far as I understand whenever `anyio.to_thread.run_sync` is gets called, it will be called with our custom limiter? 

If so this would be more readable imo:
```suggestion
        limiter = CapacityLimiter(total_tokens=self.max_threads)
        anyio.to_thread.run_sync = partial(anyio.to_thread.run_sync, limiter=limiter)
```

Reply:
But would this still hold if anyio is imported in another file again, ie `starlette` library? I am not sure this is a solution for what we're trying to do, did you try it out?

Maybe we could use this instead of `run_in_threadpool`
`await to_thread.run_sync(time.sleep, 5, limiter=limiter)
`

Reply:
> I am not sure this is a solution for what we're trying to do, did you try it out?.
I did try it out and it worked in my case. I may have overlooked other situations though.

> Maybe we could use this instead of `run_in_threadpool` `await to_thread.run_sync(time.sleep, 5, limiter=limiter)`
Sorry, I don't follow. `run_in_threadpool` goes inside starlette code and I didn't want to touch that. Why the sleep?

Reply:
Sry, my bad, missed adding this explanation:
It seems you can [spawn new threads ](https://anyio.readthedocs.io/en/stable/threads.html)in anyio, and we use `run_in_threadpool` in process_api. We might change `run_in_threadpool` with the above.

Reply:
An alternative would be to use our own `run_in_threadpool` instead of using starlette's, and provide the limiter as an additional argument.

Reply:
See [process_api](https://github.com/gradio-app/gradio/blob/main/gradio/blocks.py#L353-L356)

Reply:
> I did try it out and it worked in my case. I may have overlooked other situations though.
If it works, this design is ok as well!

Reply:
> Sry, my bad, missed adding this explanation:
> It seems you can [spawn new threads ](https://anyio.readthedocs.io/en/stable/threads.html)in anyio, and we use `run_in_threadpool` in process_api. We might change `run_in_threadpool` with the above.

Oh, you mean replace the call with our own? I'm totally fine with it!

Reply:
I'll do some testing later, in a few hours' time.